### PR TITLE
[android] NDK20 & x86_64 support

### LINKED ISSUE #617: [android] x86_64 support

### PR DESCRIPTION
**Pull request by thomasuster**
Fixes #617

- Adds support for x86_64 for android
- Ups default minSDK to 16
- Recommends NME users use  NDK20 for android
- `test` now automatically only builds the APK required for the running devices architecture

Depends on https://github.com/HaxeFoundation/hxcpp/pull/825

### CODE REVIEW

**Comment by codeservice** [comment]
And it looks like more conversions similar to if (mid > (void *)0)
needs to be done for JNI.cpp:
For example       bool Ok() const { return mMethod>(void *)0; }

**Comment by codeservice** [comment]
As I mentioned sqlite compiling well for arm7 and arm64 after replacement:
__android_log_print(ANDROID_LOG_INFO, "haxe plugin", "Got Load Proc %08x", (int)(size_t)inProc );
Issue with hxcpp/nme arm7: 
D:/HaxeToolkit/haxe/lib/hxcpp/4,0,0/include\hx/Thread.h:49:10: fatal error: 'sys/atomics.h' file not found
#include <sys/atomics.h>
         ^~~~~~~~~~~~~~~
1 error generated.

**Comment by codeservice** [comment]
#include <sys/atomics.h>
^~~~~~~~~~~~~~~
1 error generated.

Issue because I tried compile to API16 for arm. Looks like on clang I need to use API21

**Comment by codeservice** [comment]
Ok, and now can officially confirm it all can be compiled to arm7, arm64 with latest changes.
Only missing fix in hxcpp: __android_log_print(ANDROID_LOG_INFO, "haxe plugin", "Got Load Proc %08x", (int)(size_t)inProc );

**Comment by codeservice** [comment]
Is it possible compile to platform-16/armv7 on NDK20? Currently only 21 level works.
In this case I can use minimum SDK only 21. (currently 14)

More info:
https://developer.android.com/ndk/guides/application_mk#app_platform
Warning: Setting APP_PLATFORM higher than an application's minSdkVersion will likely produce an application that will not run on older devices. In most cases the libraries will fail to load because they refer to symbols that are not available on older devices.

**Comment by thomasuster** [comment]
I was able to confirm your findings, without modifying this diff I'm unable to build using the NDK platform smaller than 21. Thinking about how to proceed. 🤔

**Comment by codeservice** [comment]
Build xml has: <set name="PLATFORM" value="android-14" if="HXCPP_X86" />
   <set name="PLATFORM" value="android-21" if="HXCPP_ARM64" />
I think need to replace 14 to 16. And modify tools again to accepting that

**Comment by codeservice** [comment]
Or maybe other solution ...

**Comment by codeservice** [comment]
Also I tried this: https://github.com/HaxeFoundation/hxcpp/pull/707/files
If you add missing flag -shared it will work too.

**Comment by hughsando** [comment]
If you see my comments on the hxcpp PR, you can simply drop all the references to NDK_20, then this PR will no longer be breaking.  So, sure, if you want 86-64 you need the ndk20, but otherwise things should just work.

The jmethodId should be compared with "!=0" rather then ">"  (my bad)

Looks like the min platform for NDK20 is android-16.  If we want to go lower than that, we need an older NDK - maybe a way of side-by-side installations, and optionally retrograding via some define.  According to google, "The minimum API level supported by the latest NDK will always be low enough to support nearly all active devices" ie, android 16 is "nearly all devices". 

"If not specified, ndk-build will target the minimum API level supported by the NDK" - we could try to work out this this is done, then automatically set the platform in the hxcpp side.

Nme should probably try to default the min-sdk to the one supported by the latest compiler.  ie, once google have moved on, I guess we need to too. The logic might be best: find the latest NDK installed, or ndk explicitly specified, then find the min-platform for this NDK.  Maybe throw in an error like "You need ndk20+ for x86_64".  Optionally allow a larger platform is some java feature is required.

**Comment by codeservice** [comment]
Good news that google drop armv7 support in 2021, it will be all arm64 after. We need solution for few years only. And we need to keep in mind other libraries, not only nme. I think most using something similar to: &lt;set name="PLATFORM" value="android-14" if="HXCPP_X86" />
&lt;set name="PLATFORM" value="android-21" if="HXCPP_ARM64" /> we can ignore that and load best possible combination from tools or try satisfy requested api levels from build xml

**Comment by thomasuster** [comment]
`cd ~/nme/samples/DisplayingABitmap`
```
rm -rf ~/hxcache; rm -rf Export/android/DisplayingABitmap/app/build/outputs/apk/debug; pushd ~/hxcpp/tools/build; haxe compile.hxml; popd; pushd ~/hxcpp/tools/hxcpp; haxe compile.hxml; popd; pushd ~/nme/tools/build; haxe compile.hxml; popd; pushd ~/nme/tools/nme; haxe compile.hxml; popd; haxelib run nme build android -toolkit -gradle
```

File: project/src/android/System.cpp [hunk 1]
```diff
@@ -82,7 +82,7 @@ namespace nme {
 		JNIEnv *env = GetEnv();
       jclass cls = FindClass("org/haxe/nme/GameActivity");
 		jmethodID mid = env->GetStaticMethodID(cls, "vibrate", "(II)V");
-		if (mid > 0)
+		if (mid > (void *)0)
```
Comment (hughsando) [inline project/src/android/System.cpp]: should be "mid != 0"

**Comment by hughsando** [comment]
Great stuff, thanks!

**Pull request merged**